sa/sa.go

```diff
@@ -1549,9 +1571,164 @@ func (ssa *SQLStorageAuthority) GetOrder(ctx context.Context, req *sapb.OrderReq
 	}
 	order.Names = reversedNames
 
+	// Calculate the status for the order
+	status, err := ssa.statusForOrder(ctx, order)
+	if err != nil {
+		return nil, err
+	}
+	order.Status = &status
+
 	return order, nil
 }
 
+// statusForOrder examines the status of a provided order's authorizations to
+// determine what the overall status of the order should be. In summary:
+//   * If any of the order's authorizations are invalid, the order is invalid.
+//   * If any of the order's authorizations are deactivated, the order is deactivated.
+//   * If any of the order's authorizations are pending, the order is pending.
+//   * If all of the order's authorizations are valid, and there is
+//     a certificate serial, the order is valid.
+//   * If all of the order's authorizations are valid, and we have began
+//     processing, but there is no certificate serial, the order is processing.
+//   * If all of the order's authorizations are valid, and we haven't begun
+//     processing, then the order is pending waiting a finalization request.
+// An error is returned for any other case.
+func (ssa *SQLStorageAuthority) statusForOrder(ctx context.Context, order *corepb.Order) (string, error) {
+	// Get the full Authorization objects for the order
+	authzs, err := ssa.getAllOrderAuthorizations(ctx, *order.Id, *order.RegistrationID)
+	// If there was an error getting the authorizations, return it immediately
+	if err != nil {
+		return "", err
+	}
+
+	// If GetOrderAuthorizations returned a different number of authorization
+	// objects than the order's slice of authorization IDs something has gone
+	// wrong worth raising an internal error about.
+	if len(authzs) != len(order.Authorizations) {
+		return "", berrors.InternalServerError(
+			"GetOrderAuthorizations returned the wrong number of authorizations "+
+				"(%d vs expected %d) for order %d",
+			len(authzs), len(order.Authorizations), *order.Id)
+	}
+
+	// Keep a count of the authorizations seen
+	invalidAuthzs := 0
+	deactivatedAuthzs := 0
+	pendingAuthzs := 0
+	validAuthzs := 0
```

Comment:
Other than `validAuthz`'s there isn't a specific reason to use an `int` vs a `bool` but I thought there was something to be said for the symmetry of just counting each status type the same way. Open to feedback